{
  "info": {
    "author": "Jonas Kittner",
    "author_email": "theendlessriver13@users.noreply.github.com",
    "bugtrack_url": null,
    "classifiers": [
      "License :: OSI Approved :: MIT License",
      "Programming Language :: Python :: 3",
      "Programming Language :: Python :: 3 :: Only",
      "Programming Language :: Python :: Implementation :: CPython"
    ],
    "description": "[![ci](https://github.com/theendlessriver13/flake8-ban-utcnow/workflows/ci/badge.svg)](https://github.com/theendlessriver13/flake8-ban-utcnow/actions?query=workflow%3Aci)\n[![pre-commit.ci status](https://results.pre-commit.ci/badge/github/theendlessriver13/flake8-ban-utcnow/master.svg)](https://results.pre-commit.ci/latest/github/theendlessriver13/flake8-ban-utcnow/master)\n\n# flake8-ban-utcnow\n\nflake8 plugin which checks that `datetime.utcnow()` is not used. It suggests using `datetime.now(timezone.utc)` instead.\n\n**note:** timezone must be imported from datetime first:\n\n```python\nfrom datetime import datetime\nfrom datetime import timezone\n\ndatetime.now(timezone.utc)\n```\n\n## installation\n\n```bash\npip install flake8-ban-utcnow\n```\n\n## flake8 code\n\n| Code   | Description                                                             |\n| ------ | ----------------------------------------------------------------------- |\n| UTC001 | don't use `datetime.utcnow()`, use `datetime.now(timezone.utc)` instead |\n\n## as a pre-commit hook\n\nSee [pre-commit](https://pre-commit.com) for instructions\n\nSample `.pre-commit-config.yaml`:\n\n```yaml\n- repo: https://github.com/pycqa/flake8\n  rev: 5.0.4\n  hooks:\n    - id: flake8\n      additional_dependencies: [flake8-ban-utcnow==0.1.0]\n```\n\n## rationale\n\nOne could expect that when explicitly calling `datetime.utcnow()` the `datetime`\nobject would be timezone aware, but it's not! A common pitfall is, deriving a\ntimestamp from the `datetime` object created using `datetime.utcnow()`.\n\n### example\n\n- the computer is in `CEST` and we want to derive a `datetime` in **UTC**\n  formatted as a timestamp hence calling `utcnow().timestamp()`.\n\n  ```pycon\n  >>> from datetime import datetime\n  >>> datetime.utcnow()\n  datetime.datetime(2022, 8, 7, 23, 40, 17, 7858)\n  >>> datetime.utcnow().timestamp()\n  1659908656.048843\n  ```\n\n- if we [convert](https://www.epochconverter.com/) the timestamp, it says this,\n  which is obviously incorrect.\n\n  ```\n  GMT: Sunday, 7. August 2022 21:44:16\n  Your time zone: Sunday, 7. August 2022 23:44:16 GMT+02:00 DST\n  Relative: 2 hours ago\n  ```\n\n- converting it using python and `datetime.fromtimestamp`, we by accident get\n  the correct datetime in **UTC**\n\n  ```pycon\n  >>> datetime.fromtimestamp(1659908656.048843)\n  datetime.datetime(2022, 8, 7, 23, 44, 16, 48843)\n  ```\n\n- being aware that the timestamp _should_ be in `UTC` we call `utcfromtimestamp`\n  instead and get the result as above, since the timestamp actually is in local\n  time, but unaware of this.\n\n  ```pycon\n  >>> datetime.utcfromtimestamp(1659908656.048843)\n  datetime.datetime(2022, 8, 7, 21, 44, 16, 48843)\n  ```\n\n### the correct way\n\n- the computer is in `CEST` and we want to actually derive a `datetime` in **UTC**\n  formatted as a timestamp .\n\n  ```pycon\n  >>> from datetime import timezone\n  >>> from datetime import datetime\n  >>> datetime.now(timezone.utc).timestamp()\n  1659916399.651218\n  ```\n\n- we now get what we actually expect\n\n  ```\n  GMT: Sunday, 7. August 2022 23:53:19\n  Your time zone: Monday, 8. August 2022 01:53:19 GMT+02:00 DST\n  Relative: A few seconds ago\n  ```\n\n- the next thing to keep in mind is, that only timezone aware `datetime` objects\n  can be compared hence using this forces us to always make sure all objects are\n  timezone aware.\n",
    "description_content_type": "text/markdown",
    "docs_url": null,
    "download_url": "",
    "downloads": {
      "last_day": -1,
      "last_month": -1,
      "last_week": -1
    },
    "home_page": "https://github.com/theendlessriver13/flake8-ban-utcnow",
    "keywords": "",
    "license": "MIT",
    "maintainer": "",
    "maintainer_email": "",
    "name": "flake8-ban-utcnow",
    "package_url": "https://pypi.org/project/flake8-ban-utcnow/",
    "platform": null,
    "project_url": "https://pypi.org/project/flake8-ban-utcnow/",
    "project_urls": {
      "Homepage": "https://github.com/theendlessriver13/flake8-ban-utcnow"
    },
    "release_url": "https://pypi.org/project/flake8-ban-utcnow/0.1.0/",
    "requires_dist": [
      "flake8 (>=5.0.0)"
    ],
    "requires_python": ">=3.8",
    "summary": "flake8 plugin which bans the usage of datetime.datetime.utcnow",
    "version": "0.1.0",
    "yanked": false,
    "yanked_reason": null
  },
  "last_serial": 14683048,
  "releases": {
    "0.1.0": [
      {
        "comment_text": "",
        "digests": {
          "blake2b_256": "4328fe5678ee108a312bdc3c03c54070d0acf0ede50f3d361bdf6dd142a0297b",
          "md5": "0c97597f5af51840f5fc7cec0bee6bbb",
          "sha256": "eaa27d7a0f552ed2501a36312a6bf03ebfec71dca9f39e0d823d154cccdb009f"
        },
        "downloads": -1,
        "filename": "flake8_ban_utcnow-0.1.0-py2.py3-none-any.whl",
        "has_sig": false,
        "md5_digest": "0c97597f5af51840f5fc7cec0bee6bbb",
        "packagetype": "bdist_wheel",
        "python_version": "py2.py3",
        "requires_python": ">=3.8",
        "size": 4215,
        "upload_time": "2022-08-08T00:00:28",
        "upload_time_iso_8601": "2022-08-08T00:00:28.941326Z",
        "url": "https://files.pythonhosted.org/packages/43/28/fe5678ee108a312bdc3c03c54070d0acf0ede50f3d361bdf6dd142a0297b/flake8_ban_utcnow-0.1.0-py2.py3-none-any.whl",
        "yanked": false,
        "yanked_reason": null
      },
      {
        "comment_text": "",
        "digests": {
          "blake2b_256": "f11ebfb175f29e1359f5e2b10f41153993204ba8b8731a8a61e68c03743fcb4c",
          "md5": "d178f45a27ea32ce9f7c379013df8254",
          "sha256": "53d4c7ef638c8e248eed12f4e83d0862442c67e697a2bc0e27102a2c62841592"
        },
        "downloads": -1,
        "filename": "flake8_ban_utcnow-0.1.0.tar.gz",
        "has_sig": false,
        "md5_digest": "d178f45a27ea32ce9f7c379013df8254",
        "packagetype": "sdist",
        "python_version": "source",
        "requires_python": ">=3.8",
        "size": 3896,
        "upload_time": "2022-08-08T00:00:30",
        "upload_time_iso_8601": "2022-08-08T00:00:30.965075Z",
        "url": "https://files.pythonhosted.org/packages/f1/1e/bfb175f29e1359f5e2b10f41153993204ba8b8731a8a61e68c03743fcb4c/flake8_ban_utcnow-0.1.0.tar.gz",
        "yanked": false,
        "yanked_reason": null
      }
    ]
  },
  "urls": [
    {
      "comment_text": "",
      "digests": {
        "blake2b_256": "4328fe5678ee108a312bdc3c03c54070d0acf0ede50f3d361bdf6dd142a0297b",
        "md5": "0c97597f5af51840f5fc7cec0bee6bbb",
        "sha256": "eaa27d7a0f552ed2501a36312a6bf03ebfec71dca9f39e0d823d154cccdb009f"
      },
      "downloads": -1,
      "filename": "flake8_ban_utcnow-0.1.0-py2.py3-none-any.whl",
      "has_sig": false,
      "md5_digest": "0c97597f5af51840f5fc7cec0bee6bbb",
      "packagetype": "bdist_wheel",
      "python_version": "py2.py3",
      "requires_python": ">=3.8",
      "size": 4215,
      "upload_time": "2022-08-08T00:00:28",
      "upload_time_iso_8601": "2022-08-08T00:00:28.941326Z",
      "url": "https://files.pythonhosted.org/packages/43/28/fe5678ee108a312bdc3c03c54070d0acf0ede50f3d361bdf6dd142a0297b/flake8_ban_utcnow-0.1.0-py2.py3-none-any.whl",
      "yanked": false,
      "yanked_reason": null
    },
    {
      "comment_text": "",
      "digests": {
        "blake2b_256": "f11ebfb175f29e1359f5e2b10f41153993204ba8b8731a8a61e68c03743fcb4c",
        "md5": "d178f45a27ea32ce9f7c379013df8254",
        "sha256": "53d4c7ef638c8e248eed12f4e83d0862442c67e697a2bc0e27102a2c62841592"
      },
      "downloads": -1,
      "filename": "flake8_ban_utcnow-0.1.0.tar.gz",
      "has_sig": false,
      "md5_digest": "d178f45a27ea32ce9f7c379013df8254",
      "packagetype": "sdist",
      "python_version": "source",
      "requires_python": ">=3.8",
      "size": 3896,
      "upload_time": "2022-08-08T00:00:30",
      "upload_time_iso_8601": "2022-08-08T00:00:30.965075Z",
      "url": "https://files.pythonhosted.org/packages/f1/1e/bfb175f29e1359f5e2b10f41153993204ba8b8731a8a61e68c03743fcb4c/flake8_ban_utcnow-0.1.0.tar.gz",
      "yanked": false,
      "yanked_reason": null
    }
  ],
  "vulnerabilities": []
}